File: shared/sim/terrain.ts
import { makeValueNoise } from "./rng";
import { smoothstep } from "./math";
import { WORLD } from "./constants";

export type HeightFn = (x: number, z: number) => number;

/**
 * Build the terrain-height sampler for a seed. fBm value-noise (4 octaves) scaled by
 * hillHeight, flattened to ground level near the base camp. Ported verbatim from the
 * old Environment.getHeight so the collision mesh and the sim agree exactly.
 */
export function makeTerrain(seed: number): HeightFn {
  const noise = makeValueNoise(seed);
  return (x: number, z: number): number => {
    const nx = x * 0.0065;
    const nz = z * 0.0065;
    let h = 0;
    let amp = 1;
    let freq = 1;
    let norm = 0;
    for (let o = 0; o < 4; o++) {
      h += noise(nx * freq, nz * freq) * amp;
      norm += amp;
      amp *= 0.5;
      freq *= 2;
    }
    h = (h / norm) * WORLD.hillHeight;
    const d = Math.sqrt(x * x + z * z);
    const flat = smoothstep(d, WORLD.baseCampRadius, WORLD.baseCampRadius + 12);
    return h * flat;
  };
}

File: csharp/HollowPines.Sim/Terrain.cs
namespace HollowPines.Sim
{
    public static class Terrain
    {
        /// <summary>
        /// Build the terrain-height sampler for a seed. fBm value-noise (4 octaves) scaled by
        /// hillHeight, flattened to ground level near the base camp. Ported verbatim from
        /// shared/sim/terrain.ts so the collision mesh and the sim agree exactly.
        /// </summary>
        public static HeightFn MakeTerrain(uint seed)
        {
            var noise = Rng.MakeValueNoise(seed);
            return (x, z) =>
            {
                double nx = x * 0.0065;
                double nz = z * 0.0065;
                double h = 0;
                double amp = 1;
                double freq = 1;
                double norm = 0;
                for (int o = 0; o < 4; o++)
                {
                    h += noise(nx * freq, nz * freq) * amp;
                    norm += amp;
                    amp *= 0.5;
                    freq *= 2;
                }
                h = (h / norm) * World.HillHeight;
                double d = System.Math.Sqrt(x * x + z * z);
                double flat = SimMath.Smoothstep(d, World.BaseCampRadius, World.BaseCampRadius + 12);
                return h * flat;
            };
        }
    }
}
